File: IntersectNode.java

// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this :
//  	Not able to do intuition on this solution

// Your code here along with comments explaining your approach

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
           if (headA == null || headB == null)
        	return null;

        ListNode currA = headA, currB = headB;
        boolean isFirstRoundA = true, isFirstRoundB = true;

        while (currA != currB) {
        	currA = currA.next;
        	currB = currB.next;

        	if (currA == null && isFirstRoundA) {
        		currA = headB;
        		isFirstRoundA = false;
        	}

        	if (currB == null && isFirstRoundB) {
        		currB = headA;
        		isFirstRoundB = false;
        	}
        }

        return currA;
    }
}

File: intersectNode.py
# Time complexity O(1)
# Space complexity O(m+n)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        
        lenA, lenB = 0, 0
        
        ptr = headA
        while(ptr):
            lenA += 1
            ptr = ptr.next
        ptr = headB
        while(ptr):
            lenB += 1
            ptr = ptr.next
        
        if lenA > lenB:
            for i in range(lenA - lenB):
                headA = headA.next
        else:
            for i in range(lenB - lenA):
                headB = headB.next
        
        while(headA != headB):
            headA = headA.next
            headB = headB.next
        
        return headA